cli: TestUnavailableZip failed

cli.TestUnavailableZip [failed](https://tanzanite.cluster.engflow.com/invocations/default/c8b30906-1529-446f-bc0e-f9f80109b37f?testReportRun=3&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [34e34c2d5b9598441b935f03a794277e9329a256](https://github.com/cockroachdb/cockroach/commits/34e34c2d5b9598441b935f03a794277e9329a256):

Fatal error:

```
panic: test timed out after 25m0s
running tests:
	TestUnavailableZip (25m0s)
	TestUnavailableZip/server_2 (24m18s)

```
Stack: 

```
goroutine 47759229 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip2140135756
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip1305647791
=== RUN   TestUnavailableZip/server_1
[debug --host=127.0.0.1:46475 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
=== RUN   TestUnavailableZip/server_2
[debug --host=127.190.153.84:33639 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs @cockroachdb/server @cockroachdb/obs-india-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-47900

cli.TestUnavailableZip [failed](https://tanzanite.cluster.engflow.com/invocations/default/586a3bfa-d23b-4ab4-b6ff-54b7226d371d?testReportRun=1&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [34e34c2d5b9598441b935f03a794277e9329a256](https://github.com/cockroachdb/cockroach/commits/34e34c2d5b9598441b935f03a794277e9329a256):

Fatal error:

```
panic: test timed out after 25m0s
running tests:
	TestUnavailableZip (25m0s)
	TestUnavailableZip/server_2 (24m12s)

```
Stack: 

```
goroutine 60768728 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip1642104775
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip3389623977
=== RUN   TestUnavailableZip/server_1
[debug --host=127.0.0.1:33685 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
=== RUN   TestUnavailableZip/server_2
[debug --host=127.25.217.112:42615 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cli.TestUnavailableZip [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18859046?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18859046?buildTab=artifacts#/) on master @ [fe230d2f0b2ad9c81a7c2ebd105b5046e3746e18](https://github.com/cockroachdb/cockroach/commits/fe230d2f0b2ad9c81a7c2ebd105b5046e3746e18):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestUnavailableZip (14m7s)
	TestUnavailableZip/server_2 (13m24s)

```
Stack: 

```
goroutine 161609 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/c459704a46d4e0b3e02840af6032d657/logTestUnavailableZip4233762675
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/c459704a46d4e0b3e02840af6032d657/logTestUnavailableZip3544515013
=== RUN   TestUnavailableZip/server_2
[debug --host=127.231.177.122:35397 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cli.TestUnavailableZip [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18873806?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18873806?buildTab=artifacts#/) on master @ [b30be14916ea6a0a87df0de71fa0c429e10673fb](https://github.com/cockroachdb/cockroach/commits/b30be14916ea6a0a87df0de71fa0c429e10673fb):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestUnavailableZip (14m28s)
	TestUnavailableZip/server_2 (13m47s)

```
Stack: 

```
goroutine 221345 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/c459704a46d4e0b3e02840af6032d657/logTestUnavailableZip547507729
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/c459704a46d4e0b3e02840af6032d657/logTestUnavailableZip1647189441
=== RUN   TestUnavailableZip/server_1
[debug --host=127.0.0.1:44671 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
=== RUN   TestUnavailableZip/server_2
[debug --host=127.238.216.45:37565 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cli.TestUnavailableZip [failed](https://tanzanite.cluster.engflow.com/invocations/default/0bc5cf4d-9d51-4409-833d-a7068d1b9910?testReportRun=2&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [25ca1bb457c002a167b89f4e4e5f1a85ca315e04](https://github.com/cockroachdb/cockroach/commits/25ca1bb457c002a167b89f4e4e5f1a85ca315e04):

Fatal error:

```
panic: test timed out after 25m0s
running tests:
	TestUnavailableZip (25m0s)
	TestUnavailableZip/server_2 (24m13s)

```
Stack: 

```
goroutine 49033336 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip3996737322
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip3549671201
=== RUN   TestUnavailableZip/server_1
[debug --host=127.0.0.1:35429 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
=== RUN   TestUnavailableZip/server_2
[debug --host=127.217.25.51:45785 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cli.TestUnavailableZip [failed](https://tanzanite.cluster.engflow.com/invocations/default/38445a6e-5a86-44d5-8252-85f842916bb7?testReportRun=6&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [c471a724f99a5add3ccefe07995251760d0e2212](https://github.com/cockroachdb/cockroach/commits/c471a724f99a5add3ccefe07995251760d0e2212):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestUnavailableZip (14m25s)
	TestUnavailableZip/server_2 (13m41s)

```
Stack: 

```
goroutine 156234 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip1105459117
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip2611861942
=== RUN   TestUnavailableZip/server_2
[debug --host=127.209.106.170:44255 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cli.TestUnavailableZip [failed](https://tanzanite.cluster.engflow.com/invocations/default/ab89310d-8b5a-4fe2-b4bf-c8c03cdadeab?testReportRun=2&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [fa8342d50c1bdfbc307f4bca9bc0d34da4475258](https://github.com/cockroachdb/cockroach/commits/fa8342d50c1bdfbc307f4bca9bc0d34da4475258):

Fatal error:

```
panic: test timed out after 25m0s
running tests:
	TestUnavailableZip (25m0s)
	TestUnavailableZip/server_2 (24m5s)

```
Stack: 

```
goroutine 59176518 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip441333799
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip76189946
=== RUN   TestUnavailableZip/server_1
[debug --host=127.0.0.1:39879 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
=== RUN   TestUnavailableZip/server_2
[debug --host=127.188.159.198:33741 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

cli.TestUnavailableZip [failed](https://tanzanite.cluster.engflow.com/invocations/default/45928b17-b8d1-4b8c-92d1-7c54974593ad?testReportRun=1&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2xpOmNsaV90ZXN0) on master @ [c44b3134d23d6037e1d5d5e12461a6914aa588e3](https://github.com/cockroachdb/cockroach/commits/c44b3134d23d6037e1d5d5e12461a6914aa588e3):

Fatal error:

```
panic: test timed out after 25m0s
	running tests:
		TestUnavailableZip (25m0s)
		TestUnavailableZip/server_2 (24m18s)

```
Stack: 

```
goroutine 47604463 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2373 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:215 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestUnavailableZip
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip3998523652
    test_log_scope.go:76: use -show-logs to present logs inline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestUnavailableZip1163474478
=== RUN   TestUnavailableZip/server_2
[debug --host=127.5.173.1:33313 --insecure=true zip --concurrency=1 --cpu-profile-duration=0 /dev/null --timeout=.5s]
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnavailableZip.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/cli/zip_test.go
+++ b/pkg/cli/zip_test.go
@@ -497,8 +497,6 @@ func TestUnavailableZip(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 141565)
-
 	skip.UnderShort(t)
 	// Race builds make the servers so slow that they report spurious
 	// unavailability.

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -1273,17 +1273,8 @@ func (s *Server) newConnExecutor(
 
 	ex.extraTxnState.hasAdminRoleCache = HasAdminRoleCache{}
 
-	if lm := ex.server.cfg.LeaseManager; executorType == executorTypeExec && lm != nil {
-		if desc, err := lm.Acquire(ctx, ex.server.cfg.Clock.Now(), keys.SystemDatabaseID); err != nil {
-			log.Infof(ctx, "unable to lease system database to determine if PCR reader is in use: %s", err)
-		} else {
-			defer desc.Release(ctx)
-			// The system database ReplicatedPCRVersion is set during reader tenant bootstrap,
-			// which guarantees that all user tenant sql connections to the reader tenant will
-			// correctly set this
-			ex.isPCRReaderCatalog = desc.Underlying().(catalog.DatabaseDescriptor).GetReplicatedPCRVersion() != 0
-		}
-	}
+	// Determine if we are running on a PCR reader catalog.
+	ex.initPCRReaderCatalog(ctx)
 
 	if postSetupFn != nil {
 		postSetupFn(ex)
@@ -3850,6 +3841,34 @@ func (ex *connExecutor) initEvalCtx(ctx context.Context, evalCtx *extendedEvalCo
 		statementPreparer:    ex,
 	}
 	evalCtx.copyFromExecCfg(ex.server.cfg)
+}
+
+// initPCRReaderCatalog leases the system database to determine if
+// we are connecting to a PCR reader catalog, if this has not been attempted
+// before.
+func (ex *connExecutor) initPCRReaderCatalog(ctx context.Context) {
+	// Wait up to 10 seconds attempting to acquire the lease on the system
+	// database. Normally we should already have a lease on this object,
+	// unless there is some availability issue.
+	const initPCRReaderCatalogTimeout = 10 * time.Second
+	err := timeutil.RunWithTimeout(ctx, "detect-pcr-reader-catalog", initPCRReaderCatalogTimeout,
+		func(ctx context.Context) error {
+			if lm := ex.server.cfg.LeaseManager; ex.executorType == executorTypeExec && lm != nil {
+				desc, err := lm.Acquire(ctx, ex.server.cfg.Clock.Now(), keys.SystemDatabaseID)
+				if err != nil {
+					return err
+				}
+				defer desc.Release(ctx)
+				// The system database ReplicatedPCRVersion is set during reader tenant bootstrap,
+				// which guarantees that all user tenant sql connections to the reader tenant will
+				// correctly set this
+				ex.isPCRReaderCatalog = desc.Underlying().(catalog.DatabaseDescriptor).GetReplicatedPCRVersion() != 0
+			}
+			return nil
+		})
+	if err != nil {
+		log.Infof(ctx, "unable to lease system database to determine if PCR reader is in use: %s", err)
+	}
 }
 
 // GetPCRReaderTimestamp if the system database is setup as PCR

